### PR TITLE
Tetgen + NUMA + concurrent queue PR

### PR DESCRIPTION
**Pull request by obi1kenobi**

### CODE REVIEW

File: src/graph_compute/Makefile [hunk 1]
```diff
@@ -1,16 +1,17 @@
 CC  ?= gcc
 CXX ?= g++
-CFLAGS = -O3 -Wall
-CXXFLAGS = -fcilkplus -std=c++11 -O3 -Wall -m64
+CFLAGS = -O3 -Wall
```
Comment (obi1kenobi) [inline src/graph_compute/Makefile:3]: no space at end of line

File: src/graph_compute/bsp_scheduling.h [hunk 2]
```diff
@@ -1,7 +1,7 @@
 #ifndef BSP_SCHEDULING_H_
 #define BSP_SCHEDULING_H_
 
-#if D0_BSP
+#if D0_BSP || BASELINE
```
Comment (obi1kenobi) [inline src/graph_compute/bsp_scheduling.h]: BASELINE is not BSP, BASELINE is priority scheduling without Hilbert reordering -- see `priority_scheduling.h`

File: src/graph_compute/bsp_scheduling.h [hunk 3]
```diff
@@ -14,14 +14,18 @@ static void init_scheduling(vertex_t * const nodes, const vid_t cntNodes,
   // no-op
 }
 
-static void execute_round(const int round, vertex_t * const nodes, const vid_t cntNodes,
+static void execute_round(const int numRounds, vertex_t * const nodes,
```
Comment (obi1kenobi) [inline src/graph_compute/bsp_scheduling.h]: If we're changing the meaning of the first parameter, we should also rename the method -- `execute_round` doesn't sound like it executes all rounds.

File: src/graph_compute/chunk_scheduling.h [hunk 4]
```diff
@@ -28,6 +28,8 @@ static inline bool interChunkDependency(vid_t v, vid_t w) {
   static const vid_t chunkMask = (1 << CHUNK_BITS) - 1;
   if ((v >> CHUNK_BITS) == (w >> CHUNK_BITS)) {
     return false;
+  } else if ((v & chunkMask) == (w & chunkMask)) {
```
Comment (obi1kenobi) [inline src/graph_compute/chunk_scheduling.h:34]: `(v & chunkMask) == (w & chunkMask)` is only true if `v` and `w` are in the same chunk, no? Why can this return true if `v` depends on `w` but in the same chunk? It's confusing...

File: src/graph_compute/compute.cpp [hunk 5]
```diff
@@ -12,46 +12,69 @@
 #include "./common.h"
 #include "./io.h"
 #include "./update_function.h"
+#include "./concurrent_queue.h"
 
 using namespace std;
 
-#if BASELINE || D1_PRIO
-  #include "./priority_scheduling.h"
+#if BASELINE || D0_BSP
```
Comment (obi1kenobi) [inline src/graph_compute/compute.cpp]: Why are we changing what BASELINE refers to? This seems unnecessary and confusing.

File: src/graph_compute/phase_scheduling.h [hunk 6]
```diff
@@ -40,6 +40,8 @@ static inline bool interChunkDependency(vid_t v, vid_t w) {
   static const vid_t chunkMask = (1 << CHUNK_BITS) - 1;
   if ((v >> CHUNK_BITS) == (w >> CHUNK_BITS)) {
     return false;
+  } else if ((v & chunkMask) == (w & chunkMask)) {
```
Comment (obi1kenobi) [inline src/graph_compute/phase_scheduling.h:55]: Again, this doesn't look like a valid inter-chunk dependency -- why can this return true?

File: src/graph_compute/common.h [hunk 7]
```diff
@@ -75,17 +92,6 @@
...
-  vid_t id;
-  vid_t priority;
-  vid_t dependencies;
-  volatile vid_t satisfied;
-  vid_t cntEdges;
-  vid_t * edges;
-  double data;
-};
-typedef struct vertex_t vertex_t;
+//  typedef uint32_t vid_t;  // vertex id type
```
Comment (obi1kenobi) [inline src/graph_compute/common.h]: `vid_t` has no definition -- this one is commented out, and there is no other one. Also, why did the type change from `uint64_t` to `uint32_t`?

File: src/graph_compute/compute.cpp [hunk 8]
```diff
@@ -12,46 +12,69 @@
...
 #endif
 
-WHEN_TEST(
-  static uint64_t roundUpdateCount = 0;
-)
-
 // fill in each node with random-looking data
 static void fillInNodeData(vertex_t * nodes, const vid_t cntNodes) {
-  cilk_for (vid_t i = 0; i < cntNodes; ++i) {
+  for (vid_t i = 0; i < cntNodes; ++i) {
```
Comment (obi1kenobi) [inline src/graph_compute/compute.cpp]: Why remove the cilk_for?

File: src/graph_compute/numa_init.cpp [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+    assert(pthread_create(&workers[i], NULL, writeZeroes, &chunk) == 0);
+  }
+  for (int i = 0; i < config.numWorkers; i++) {
+    assert(pthread_join(workers[i], NULL) == 0);
+  }
+}
+
+void * numaCalloc(numaInit_t config, size_t dataTypeSize, size_t numElements) {
+  void * data = malloc(numElements*dataTypeSize);
+  madvise(data, numElements*dataTypeSize, MADV_NOHUGEPAGE);
```
Comment (obi1kenobi) [inline src/graph_compute/numa_init.cpp:54]: Why do we want `MADV_NOHUGEPAGE`?

File: src/graph_compute/numa_init.h [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+#include <sched.h>
+#include <unistd.h>
+#include <pthread.h>
+#include <sys/mman.h>
+#include <string>
+#include <cstdio>
+#include <cassert>
+#include <cstring>
+#include <algorithm>
+// #include "./io.h"
```
Comment (obi1kenobi) [inline src/graph_compute/numa_init.h]: If these headers aren't needed, please delete them.

File: src/graph_compute/update_function.h [hunk 11]
```diff
@@ -1,6 +1,86 @@
...
+    // ensuring that the number of updates in total is correct per round
+    WHEN_TEST({
+      __sync_add_and_fetch(&roundUpdateCount, 1);
+    })
+
+    vertex_t * current = &nodes[index];
+    // recalculate this node's data
+    double data = current->data;
+    for (size_t i = 0; i < current->cntEdges; ++i) {
+      data += nodes[current->edges[i]].data;
```
Comment (obi1kenobi) [inline src/graph_compute/update_function.h]: This doesn't seem like the right update function for mass spring dashpot.

File: src/graph_compute/Makefile (hunk 1 above)
Comment (whasenpl) [inline src/graph_compute/Makefile:3]: Fixed in my current branch 'specialized_vertex_t' , which is ahead of this one.

File: src/graph_compute/bsp_scheduling.h (hunk 2 above)
Comment (whasenpl) [inline src/graph_compute/bsp_scheduling.h]: I changed it - BSP is a more natural baseline.

File: src/graph_compute/bsp_scheduling.h (hunk 3 above)
Comment (whasenpl) [inline src/graph_compute/bsp_scheduling.h]: Changed it to 'execute_rounds.'

File: src/graph_compute/bsp_scheduling.h (hunk 2 above)
Comment (obi1kenobi) [inline src/graph_compute/bsp_scheduling.h]: I understand, but this is more than just a naming change -- BASELINE is a run configuration that the scripts depend on and use. If we want to compare against D0_BSP, we can just use the D0_BSP flag. If you make BASELINE mean the same thing as D0_BSP, then there will be no way to run what used to be the BASELINE configuration.

File: src/graph_compute/chunk_scheduling.h (hunk 4 above)
Comment (whasenpl) [inline src/graph_compute/chunk_scheduling.h:34]: No, ((v & chunkMask) == (w & chunkMask)) implies that they are at the same position within their respective chunks (recall: the previous if which failed, checked if they are in the same chunk).

File: src/graph_compute/bsp_scheduling.h (hunk 3 above)
Comment (obi1kenobi) [inline src/graph_compute/bsp_scheduling.h]: Can you change it in this PR? Just push to the `pr_rollback` branch.

File: src/graph_compute/compute.cpp (hunk 5 above)
Comment (whasenpl) [inline src/graph_compute/compute.cpp]: I don't think it's all that confusing.  Admittedly, it isn't all that necessary, either.  I did it anyway.  Probably not the most unnecessary or confusing change, actually.

File: src/graph_compute/chunk_scheduling.h (hunk 4 above)
Comment (obi1kenobi) [inline src/graph_compute/chunk_scheduling.h:34]: So you are saying this was a bug in the original implementation?

File: src/graph_compute/phase_scheduling.h (hunk 6 above)
Comment (whasenpl) [inline src/graph_compute/phase_scheduling.h:55]: It returns true if v is at a lower position within its chunk than w is within its chunk.  If they are at the same position, within their respective chunks, we break ties by the chunk number.

File: src/graph_compute/compute.cpp (hunk 5 above)
Comment (obi1kenobi) [inline src/graph_compute/compute.cpp]: See my reply on the other BASELINE comment.

File: src/graph_compute/common.h (hunk 7 above)
Comment (whasenpl) [inline src/graph_compute/common.h]: I moved the vertex_t definition to update_function.h, in order to facilitate it's composition of three pieces, as I described in the recent email:
a sched_t which is defined by each scheduling algorithm
a data_t which is defined by each application (e.g., pagerank and mass-spring-dashpot)
and vid_t \* edges, vid_t cntEdges, used by everyone

The vid_t definition is in libgraphio.h

Comment (whasenpl) [inline src/graph_compute/common.h]: oh, I had contemplated changing vid_t to a uint32_t, but it broke a bunch of code in the graph loading library which has constants in read routines which are specific to 8-byte ints.

Comment (obi1kenobi) [inline src/graph_compute/common.h]: Can you remove this line then?

File: src/graph_compute/compute.cpp (hunk 8 above)
Comment (whasenpl) [inline src/graph_compute/compute.cpp]: I want to remove cilk from all but the timed scheduling-specific routines, so that there are no issues with the creation of pthreads in the numa_scheduling code.

Comment (obi1kenobi) [inline src/graph_compute/compute.cpp]: I thought you might be doing that, but then I saw `test_queue` using cilk and I was confused...

File: src/graph_compute/numa_init.cpp (hunk 9 above)
Comment (whasenpl) [inline src/graph_compute/numa_init.cpp:54]: This is a placeholder - Bradley told me that with numa-based allocations, huge pages can sometimes cause problems.  I got this call from him and I thought I'd test it with and without.

File: src/graph_compute/numa_init.h (hunk 10 above)
Comment (whasenpl) [inline src/graph_compute/numa_init.h]: Done - I deleted the unnecessary ~half of them.

File: src/graph_compute/update_function.h (hunk 11 above)
Comment (whasenpl) [inline src/graph_compute/update_function.h]: It's not.  My current branch has the correct update function for it, minus the code to load the positions from a *.nodes file.

File: src/graph_compute/numa_init.h (hunk 10 above)
Comment (obi1kenobi) [inline src/graph_compute/numa_init.h]: Can you update this pull request with the change? You need to commit and push to its branch, `pr_rollback`.

File: src/graph_compute/bsp_scheduling.h (hunk 2 above)
Comment (whasenpl) [inline src/graph_compute/bsp_scheduling.h]: What else does BASELINE imply, other than the scheduling algorithm?

Comment (obi1kenobi) [inline src/graph_compute/bsp_scheduling.h]: BASELINE is priority scheduling executed without Hilbert reordering. Open any of the benchmarking scripts and you'll see that's how they use it. If you change the definition of BASELINE, all of those scripts will stop giving meaningful results, which is why I don't think this is worth it and I'd prefer it if it got changed back to what it was.

File: src/graph_compute/bsp_scheduling.h (hunk 3 above)
Comment (whasenpl) [inline src/graph_compute/bsp_scheduling.h]: I don't understand - do you want me to check out pr_rollback, make this change and push it to advance pr_rollback?  I have a branch, specialized_vertex_t, that started off as the current pr_rollback and has evolved significantly since then.  The comments you are making here are quite relevant to this new branch, which is why I'm making the changes there, rather than pr_rollback.  I was expecting that after addressing your concerns I would do a pull request vs pr_rollback.  How should this process work?

File: src/graph_compute/chunk_scheduling.h (hunk 4 above)
Comment (whasenpl) [inline src/graph_compute/chunk_scheduling.h:34]: Oh, that's chunk_scheduling.h, not phase_scheduling.h.  Sorry.  chunk_scheduling.h is essentially abandoned, as it is totally dominated by phase_scheduling.h.  In fact, I think if you just set NUM_PHASES=1, it becomes chunk_scheduling.h.  I'm not sure about the chunk_scheduling.h logic, the logic in phase_scheduling.h looks correct to me.

File: src/graph_compute/common.h (hunk 7 above)
Comment (whasenpl) [inline src/graph_compute/common.h]: Done.

File: src/graph_compute/compute.cpp (hunk 8 above)
Comment (whasenpl) [inline src/graph_compute/compute.cpp]: I guess test_queue should do a pthreads version, but I was being lazy and wanted to rush to see if concurrent_queue.h worked.

File: src/graph_compute/bsp_scheduling.h (hunk 3 above)
Comment (obi1kenobi) [inline src/graph_compute/bsp_scheduling.h]: Generally, the idea is to not merge pull requests until all comments are resolved, so that the master branch always has a working, satisfactory version of the code. That would mean advancing `pr_rollback` by pushing to it, and rebasing other branches as necessary.

This is called the feature-branch workflow, see here for details: https://www.atlassian.com/git/tutorials/comparing-workflows/feature-branch-workflow

File: src/graph_compute/bsp_scheduling.h (hunk 2 above)
Comment (whasenpl) [inline src/graph_compute/bsp_scheduling.h]: OK - that's fair.

File: src/graph_compute/bsp_scheduling.h (hunk 3 above)
Comment (whasenpl) [inline src/graph_compute/bsp_scheduling.h]: So, does this not imply that I would have to merge in these changes to my current branch?  That seems like a pain in the ass.  Of course, if we serialize the process: branch, do work, get feedback & address it, commit to master, repeat.  This seems nice and neat, however, I'm in an insane hurry to get this working and serializing worries me...

File: src/graph_compute/bsp_scheduling.h (hunk 2 above)
Comment (whasenpl) [inline src/graph_compute/bsp_scheduling.h]: Done.

File: src/graph_compute/phase_scheduling.h (hunk 6 above)
Comment (obi1kenobi) [inline src/graph_compute/phase_scheduling.h:55]: So was the lack of this line a bug here originally? You didn't answer my question in the other comment.

File: src/graph_compute/bsp_scheduling.h (hunk 3 above)
Comment (obi1kenobi) [inline src/graph_compute/bsp_scheduling.h]: Not merge, rebase -- very different. If you've made a lot of progress since the latest commit in this PR, then update this PR with all of your branch's changes -- ideally with a rebase but a merge should also work.

File: src/graph_compute/phase_scheduling.h (hunk 6 above)
Comment (whasenpl) [inline src/graph_compute/phase_scheduling.h:55]: Oh, that's weird - I haven't touched this code in quite a while.  I'm not sure why master had this bug.  It looks like it got fixed when I committed the base meshes, so I must have had that bug fix in my local copy.

File: src/graph_compute/bsp_scheduling.h [hunk 12]
```diff
@@ -4,36 +4,45 @@
...
+
+    cilk_for (vid_t i = 0; i < cntNodes; ++i) {
+      update(nodes, i, globaldata, round);
+    }
   }
 }
 
-static void cleanup_scheduling(vertex_t * const nodes, const vid_t cntNodes,
+static inline void cleanup_scheduling(vertex_t * const nodes, const vid_t cntNodes,
                                scheddata_t * const scheddata) {
```
Comment (obi1kenobi) [inline src/graph_compute/bsp_scheduling.h]: Match the indentation of the args in the row above.

File: src/graph_compute/bsp_scheduling.h [hunk 13]
```diff
@@ -4,36 +4,45 @@
...
   // no-op
 }
 
-static void print_execution_data() {
+static inline void print_execution_data() {
   // no-op
 }
 
-#endif  // D0_BSP
+#endif  // D0_BSP || BASELINE
```
Comment (obi1kenobi) [inline src/graph_compute/bsp_scheduling.h]: Update the comment to match the `#if`

File: src/graph_compute/concurrent_queue.h [hunk 14]
```diff
@@ -0,0 +1,48 @@
+#ifndef CONCURRENT_QUEUE_H_
+#define CONCURRENT_QUEUE_H_
+
+#include "../libgraphio/libgraphio.h"
+
+struct mrmw_queue_t {
```
Comment (obi1kenobi) [inline src/graph_compute/concurrent_queue.h:6]: It would be nice to walk through this code in person in the next meeting, it's very complex to read in isolation.

File: src/graph_compute/io.cpp [hunk 15]
```diff
@@ -58,9 +65,9 @@ class ComputeEdgeListBuilder : public EdgeListBuilder {
   }
 
   void build() {
-    this->nodes[0].id = 0;
+    //  this->nodes[0].id = 0;
```
Comment (obi1kenobi) [inline src/graph_compute/io.cpp]: Please don't leave commented code in. If it's not used, it should be deleted.

File: src/graph_compute/numa_init.cpp [hunk 16]
```diff
@@ -0,0 +1,61 @@
+#include "./numa_init.h"
+#include <iostream>
+
+// core_id = 0, 1, ... n-1, where n is the system's number of cores
+
+int bindThreadToCore(int _coreID) {
```
Comment (obi1kenobi) [inline src/graph_compute/numa_init.cpp:6]: We should also go through this code in person next time.

File: src/graph_compute/phase_scheduling.h [hunk 17]
```diff
@@ -20,13 +19,23 @@ typedef struct chunkdata_t chunkdata_t;
 
 struct scheddata_t {
   vid_t * dependentEdges;
-  vid_t * dependentEdgeIndex;
-  vid_t * cntDependentEdges;
+//  vid_t * dependentEdgeIndex;
```
Comment (obi1kenobi) [inline src/graph_compute/phase_scheduling.h]: Please don't leave commented code in.

File: src/graph_compute/phase_scheduling.h [hunk 18]
```diff
@@ -20,13 +19,23 @@ typedef struct chunkdata_t chunkdata_t;
 
 struct scheddata_t {
   vid_t * dependentEdges;
```
Comment (obi1kenobi) [inline src/graph_compute/phase_scheduling.h]: Why does `dependentEdges` exist both in `scheddata_t` and in `sched_t`?

File: src/graph_compute/phase_scheduling.h [hunk 19]
```diff
@@ -20,13 +19,23 @@ typedef struct chunkdata_t chunkdata_t;
...
+//  vid_t * cntDependentEdges;
   chunkdata_t * chunkdata;
   vid_t cntChunks;
 };
 typedef struct scheddata_t scheddata_t;
 
+struct sched_t {
+  vid_t * dependentEdges;
+  vid_t cntDependentEdges;
+  vid_t dependencies;
```
Comment (obi1kenobi) [inline src/graph_compute/phase_scheduling.h]: Should this be a pointer?

File: src/graph_compute/phase_scheduling.h [hunk 20]
```diff
@@ -40,12 +49,14 @@ static inline bool interChunkDependency(vid_t v, vid_t w) {
...
+  } else if ((v & chunkMask) == (w & chunkMask)) {
+    return (v < w);
   } else {
     return ((v & chunkMask) < (w & chunkMask));
   }
 }
 
-static void calculateNeighborhood(std::unordered_set<vid_t> * neighbors,
+static inline void calculateNeighborhood(std::unordered_set<vid_t> * neighbors,
                                   std::unordered_set<vid_t> * oldNeighbors,
```
Comment (obi1kenobi) [inline src/graph_compute/phase_scheduling.h]: Please align parameters to the row above.

File: src/graph_compute/phase_scheduling.h [hunk 21]
```diff
@@ -64,11 +75,10 @@ static void calculateNeighborhood(std::unordered_set<vid_t> * neighbors,
   }
 }
 
-static void calculateNodeDependenciesChunk(vertex_t * const nodes,
+static inline void calculateNodeDependenciesChunk(vertex_t * const nodes,
                                            const vid_t cntNodes,
```
Comment (obi1kenobi) [inline src/graph_compute/phase_scheduling.h]: Please align parameters.

File: src/graph_compute/priority_scheduling.h [hunk 22]
```diff
@@ -173,30 +193,31 @@ static void init_scheduling(vertex_t * const nodes, const vid_t cntNodes,
...
   delete[] scheddata->roots;
 }
 
-static void print_execution_data() {
+static inline void print_execution_data() {
   cout << "Priority group bits: " << PRIORITY_GROUP_BITS << '\n';
 }
 
-#endif  // D1_PRIO || BASELINE
+#endif  // D1_PRIO
```
Comment (obi1kenobi) [inline src/graph_compute/priority_scheduling.h]: Update comment to match `#if` above.

File: src/graph_compute/priority_scheduling.h [hunk 23]
```diff
@@ -173,30 +193,31 @@ static void init_scheduling(vertex_t * const nodes, const vid_t cntNodes,
...
+    })
+    cilk_for (vid_t i = 0; i < scheddata->cntRoots; ++i) {
+      processNode(nodes, scheddata->roots[i], cntNodes, 0, globaldata, round);
+    }
   }
 }
 
-static void cleanup_scheduling(vertex_t * const nodes, const vid_t cntNodes,
+static inline void cleanup_scheduling(vertex_t * const nodes, const vid_t cntNodes,
                                scheddata_t * const scheddata) {
```
Comment (obi1kenobi) [inline src/graph_compute/priority_scheduling.h]: Align parameters to row above.

File: src/graph_compute/priority_scheduling.h [hunk 24]
```diff
@@ -157,13 +177,13 @@ static void findRoots(vertex_t * const nodes, const vid_t cntNodes,
...
-    if (nodes[i].dependencies == 0) {
+    if (nodes[i].sched.dependencies == 0) {
       scheddata->roots[position++] = i;
     }
   }
 }
 
-static void init_scheduling(vertex_t * const nodes, const vid_t cntNodes,
+static inline void init_scheduling(vertex_t * const nodes, const vid_t cntNodes,
                             scheddata_t * const scheddata) {
```
Comment (obi1kenobi) [inline src/graph_compute/priority_scheduling.h]: Align parameters to row above.

File: src/graph_compute/priority_scheduling.h [hunk 25]
```diff
@@ -113,10 +131,10 @@ static inline id_t createPriority(const vid_t id, const int bitsInId) {
   return priority;
 }
 
-static void assignNodePriorities(vertex_t * const nodes, const vid_t cntNodes,
+static inline void assignNodePriorities(vertex_t * const nodes, const vid_t cntNodes,
                                  const int bitsInId) {
```
Comment (obi1kenobi) [inline src/graph_compute/priority_scheduling.h]: Align parameters to row above.

File: src/graph_compute/phase_scheduling.h [hunk 26]
```diff
@@ -125,66 +136,69 @@ static void createChunkData(vertex_t * const nodes, const vid_t cntNodes,
...
-          scheddata->chunkdata[i].nextIndex = j;
         }
       }
     }
   }
 }
 
-static void cleanup_scheduling(vertex_t * const nodes, const vid_t cntNodes,
+static inline void cleanup_scheduling(vertex_t * const nodes, const vid_t cntNodes,
                                scheddata_t * const scheddata) {
```
Comment (obi1kenobi) [inline src/graph_compute/phase_scheduling.h]: Align parameters to row above.

File: src/graph_compute/phase_scheduling.h [hunk 27]
```diff
@@ -125,66 +136,69 @@ static void createChunkData(vertex_t * const nodes, const vid_t cntNodes,
   }
 }
 
-static void init_scheduling(vertex_t * const nodes, const vid_t cntNodes,
+static inline void init_scheduling(vertex_t * const nodes, const vid_t cntNodes,
                             scheddata_t * const scheddata) {
```
Comment (obi1kenobi) [inline src/graph_compute/phase_scheduling.h]: Align parameters.

File: src/graph_compute/phase_scheduling.h [hunk 28]
```diff
@@ -117,7 +128,7 @@ static void createChunkData(vertex_t * const nodes, const vid_t cntNodes,
   cilk_for (vid_t i = 0; i < scheddata->cntChunks; ++i) {
     chunkdata_t * chunk = &scheddata->chunkdata[i];
     chunk->nextIndex = i << CHUNK_BITS;
-    chunk->phaseEndIndex[0] = std::min(chunk->nextIndex + (1 << (CHUNK_BITS - 1)),
+    chunk->phaseEndIndex[0] = std::min(chunk->nextIndex + ((1 << CHUNK_BITS) >> 1),
```
Comment (obi1kenobi) [inline src/graph_compute/phase_scheduling.h]: Why the change?

File: src/graph_compute/phase_scheduling.h [hunk 29]
```diff
@@ -76,39 +86,40 @@ static void calculateNodeDependenciesChunk(vertex_t * const nodes,
...
+        && (interChunkDependency(i, neighbor))) {
+          sched->dependentEdges[curIndex++] = neighbor;
       }
     }
   }
 }
 
-static void createChunkData(vertex_t * const nodes, const vid_t cntNodes,
+static inline void createChunkData(vertex_t * const nodes, const vid_t cntNodes,
                             scheddata_t * const scheddata) {
```
Comment (obi1kenobi) [inline src/graph_compute/phase_scheduling.h]: Align parameters.

File: src/graph_compute/phase_scheduling.h (hunk 19 above)
Comment (obi1kenobi) [inline src/graph_compute/phase_scheduling.h:35]: Would be nice to add comments explaining what each of these fields does, for the sake of other people reviewing our code.

File: src/graph_compute/phase_scheduling.h (hunk 28 above)
Comment (whasenpl) [inline src/graph_compute/phase_scheduling.h]: CHUNK_BITS can be 0, which generates a compilation error.  The new version doesn't generate an error and is equivalent in meaning when CHUNK_BITS is greater than 0.

File: src/graph_compute/bsp_scheduling.h (hunk 12 above)
Comment (whasenpl) [inline src/graph_compute/bsp_scheduling.h]: Done.

File: src/graph_compute/bsp_scheduling.h (hunk 13 above)
Comment (whasenpl) [inline src/graph_compute/bsp_scheduling.h]: Done

File: src/graph_compute/io.cpp (hunk 15 above)
Comment (whasenpl) [inline src/graph_compute/io.cpp]: Done.

File: src/graph_compute/phase_scheduling.h (hunk 17 above)
Comment (whasenpl) [inline src/graph_compute/phase_scheduling.h]: Done

File: src/graph_compute/phase_scheduling.h (hunk 18 above)
Comment (whasenpl) [inline src/graph_compute/phase_scheduling.h]: The one in scheddata holds the root of the dependent edges array, whereas the one in the sched_t points into that array on a per-vertex basis.

File: src/graph_compute/phase_scheduling.h (hunk 20 above)
Comment (whasenpl) [inline src/graph_compute/phase_scheduling.h]: Done

File: src/graph_compute/phase_scheduling.h (hunk 21 above)
Comment (whasenpl) [inline src/graph_compute/phase_scheduling.h]: done

File: src/graph_compute/priority_scheduling.h (hunk 22 above)
Comment (whasenpl) [inline src/graph_compute/priority_scheduling.h]: done

File: src/graph_compute/priority_scheduling.h (hunk 23 above)
Comment (whasenpl) [inline src/graph_compute/priority_scheduling.h]: done

File: src/graph_compute/priority_scheduling.h (hunk 24 above)
Comment (whasenpl) [inline src/graph_compute/priority_scheduling.h]: done

File: src/graph_compute/priority_scheduling.h (hunk 25 above)
Comment (whasenpl) [inline src/graph_compute/priority_scheduling.h]: done

File: src/graph_compute/phase_scheduling.h (hunk 26 above)
Comment (whasenpl) [inline src/graph_compute/phase_scheduling.h]: done

File: src/graph_compute/phase_scheduling.h (hunk 27 above)
Comment (whasenpl) [inline src/graph_compute/phase_scheduling.h]: done

File: src/graph_compute/phase_scheduling.h (hunk 29 above)
Comment (whasenpl) [inline src/graph_compute/phase_scheduling.h]: done

File: src/graph_compute/phase_scheduling.h (hunk 19 above)
Comment (whasenpl) [inline src/graph_compute/phase_scheduling.h:35]: done

File: src/graph_compute/phase_scheduling.h (hunk 28 above)
Comment (obi1kenobi) [inline src/graph_compute/phase_scheduling.h]: What does setting CHUNK_BITS to zero mean? Shouldn't that be a nonsensical setting that should generate a compilation error? 

—

Comment (whasenpl) [inline src/graph_compute/phase_scheduling.h]: True, though I don't think this is where the error should come from (i.e., some technical calculation deep in the initialization code).  Instead, we should probably have a preprocessor error at the top of phase_scheduler.h.

Comment (obi1kenobi) [inline src/graph_compute/phase_scheduling.h]: In that case, can you revert this change and add a preprocessor error like you mentioned? 

—

Comment (obi1kenobi) [inline src/graph_compute/phase_scheduling.h]: I think this is the last thing -- everything else looks good, and I'll merge the PR into master as soon as you get a chance to fix this.

Comment (whasenpl) [inline src/graph_compute/phase_scheduling.h]: OK - I pushed the the changes to pr_rollback.  I even learned how to use stash to go back, apply this change, then rebase the numa_scheduling branch, and finally apply the stashed changes.  Of course, it was a pain in the ass and I only knew what was going on about 40% of the time, but it seems to have worked.  Anyway, I'll do a PR for numa_scheduling relative to master tomorrow sometime.
